packages/sd-jwt/src/SdJwtService.ts

```diff
@@ -0,0 +1,179 @@
+    this.logger = logger
+  }
+
+  private hasher(input: string) {
+    const serializedInput = TypedArrayEncoder.fromString(input)
+    const hash = Hasher.hash(serializedInput, 'sha2-256')
+
+    return TypedArrayEncoder.toBase64URL(hash)
+  }
+
+  /**
+   * @todo validate the JWT header (alg)
+   */
+  private signer<Header extends Record<string, unknown> = Record<string, unknown>>(
+    agentContext: AgentContext,
+    key: Key
+  ): Signer<Header> {
+    return async (input: string) => agentContext.wallet.sign({ key, data: TypedArrayEncoder.fromString(input) })
+  }
+
+  /**
+   * @todo validate the JWT header (alg)
+   */
+  private verifier<Header extends Record<string, unknown> = Record<string, unknown>>(
+    agentContext: AgentContext,
+    issuerKey: Key
+  ): Verifier<Header> {
+    return async ({ message, signature }) => {
+      return await agentContext.wallet.verify({
+        signature: Buffer.from(signature),
+        key: issuerKey,
+        data: TypedArrayEncoder.fromString(message),
+      })
+    }
+  }
+
+  public async create<Payload extends Record<string, unknown> = Record<string, unknown>>(
+    agentContext: AgentContext,
+    jwt: Jwt | Payload,
+    { issuerKey, disclosureFrame, hashingAlgorithm = 'sha2-256', holderKey }: SdJwtCreateOptions<Payload>
+  ): Promise<SdJwtRecord> {
+    if (hashingAlgorithm !== 'sha2-256') {
+      throw new SdJwtError(`Unsupported hashing algorithm used: ${hashingAlgorithm}`)
+    }
+
+    const { header, payload } =
+      jwt instanceof Jwt
+        ? { header: jwt.header, payload: jwt.payload.toJson() }
+        : {
+            header: { alg: getJwaFromKeyType(issuerKey.keyType).toString() },
+            payload: jwt,
+          }
+
+    const confirmationClaim = holderKey ? getJwkFromKey(holderKey).toJson() : undefined
+
+    let sdJwt = new SdJwt<typeof header, typeof payload>()
+      .withHeader(header)
+      .withPayload(payload)
+      .withHasher({ hasher: this.hasher, algorithm: HasherAlgorithm.Sha256 })
+      .withSigner(this.signer(agentContext, issuerKey))
+      .withSaltGenerator(agentContext.wallet.generateNonce)
+
+    sdJwt = confirmationClaim ? sdJwt.addPayloadClaim('cnf', confirmationClaim) : sdJwt
+    sdJwt = disclosureFrame ? sdJwt.withDisclosureFrame(disclosureFrame) : sdJwt
+
+    const compact = await sdJwt.toCompact()
+
+    return new SdJwtRecord({
+      sdJwt: compact,
+    })
+  }
+
+  /**
+   * @todo Name is not the best
+   * @todo fix with the newer API
+   */
+  public async receive(
+    agentContext: AgentContext,
+    sdJwt: SdJwt,
+    { holderKey, issuerKey }: SdJwtReceiveOptions
+  ): Promise<SdJwtRecord> {
+    const isValid = await sdJwt.verifySignature(this.verifier(agentContext, issuerKey))
+
+    if (!isValid) {
+      throw new SdJwtError(`sd-jwt is not valid.`)
+    }
+
+    // TODO: append holder key here
+    const compact = await sdJwt.toCompact()
+
+    return new SdJwtRecord({
+      sdJwt: compact,
+    })
```

Comment:
Are we interested in supporting generic SD-JWTs? Wouldn't we rather want to support the VC variant of SD-JWT and be able to re-use the W3C Credential record for this with a new format?

Reply:
Tbh I would not like to restrict the user here. Might need to have a short discussion on the relation between the w3c service and the sdjwt service.

Reply:
We discussed this a bit, and right now (as a start) we will only support sd-jwt-vc. Now the question is whether we make the module 'sd-jwt-vc' / 'SdJwtVcModule' or leave it as sd-jwt.

Reply:
I think it should become a service in the W3cCredentialsModule then. We already have services for JsonLd and Jwt, we can add the SD part to it? Or maybe it becomes a completely separate module and we call it SdJwtCredentialsModule (to make it consistent with W3cCredentials?)

Reply:
sd-jwt-vc is compatible with the W3c spec though right? You're saying we will only support `sd-jwt-vc`

Reply:
sd-jwt-vc is not compatible with the w3c data model specs.

Reply:
Ok, makes sense then 👍